### PR TITLE
Migrate muting contacts using form

### PR DESCRIPTION
**Pull request by tatilepizs**
# Description

Migrate the forms-against-muted-contacts test from protractor to wdio.

This test:

- Edit the config settings to add the options for `muting`.
   ```
    muting: {
      mute_forms: ['death_report'],
      unmute_forms: ['undo_death_report']
    }
   ```
- Verify that the contact was muted after when the `death_report` is submitted.
- Verify that the contact was unmuted after when the `undo_death_report` is submitted.
- Verify that a popup is displayed when trying to submit a non-unmuting form against a muted contact.

cc: @andrablaj 

# Code review checklist

- [ ] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [ ] Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- [ ] Tested: Unit and/or e2e where appropriate
- [ ] Internationalised: All user facing text
- [ ] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# Compose URLs

If Build CI hasn't passed, these may 404:

* [Core](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:migrate-muting-contacts-using-form/docker-compose/cht-core.yml)
* [CouchDB Single](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:migrate-muting-contacts-using-form/docker-compose/cht-couchdb.yml)
* [CouchDB Cluster](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:migrate-muting-contacts-using-form/docker-compose/cht-couchdb-clustered.yml)
 
# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

File: tests/page-objects/default/contacts/contacts.wdio.page.js [hunk 1]
```diff
@@ -235,21 +236,23 @@ const editDistrict = async (districtName, editedName) => {
...
-const openForm = async (name) => {
+const openForm = async (name, waitForDisplayed = true) => {
   const parent = await newActionContactButton().parentElement();
   await browser.waitUntil(async () => await parent.getAttribute('aria-expanded') === 'true');
 
   for (const form of await forms()) {
     if (await form.getText() === name) {
       await form.click();
-      await (await formTitle()).waitForDisplayed();
+      if(waitForDisplayed) {
```
Comment (tatilepizs) [inline tests/page-objects/default/contacts/contacts.wdio.page.js:253]: I am sure there are better ways to manage this, but this is the only idea that I had.
Any suggestions?

**Comment by tatilepizs** [comment]
@latin-panda @lorerod  👋
Could you please help me with the review of this e2e test?
Thanks!

File: tests/page-objects/default/contacts/contacts.wdio.page.js (hunk 1 above)
Comment (latin-panda) [reply tests/page-objects/default/contacts/contacts.wdio.page.js:253]: Some options come to mind:
1. Change the parameter name to `shouldDisplayTitle` or `shouldDisplayContent` 
2. Don't ever check the content here and do that outside this function
3. Create another function (`openFormWithWarning`) that verifies a modal opens and returns the modal's content. So, the spec is in charge of checking the modal's message. 

I think the parameter `waitForDisplayed` isn't describing what is happening.

**Review by latin-panda — CHANGES_REQUESTED** [review]
Nice job!

File: tests/page-objects/default/contacts/contacts.wdio.page.js (hunk 1 above)
Comment (tatilepizs) [reply tests/page-objects/default/contacts/contacts.wdio.page.js]: Thanks Jenny.
I created the new function `openFormWithWarning` 😄

**Review by latin-panda — APPROVED** [review]
Good job!✨✨

File: tests/e2e/default/contacts/muting-contact-using-form.wdio-spec.js [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+const deathReportForm = require('../../../page-objects/default/enketo/death-report.page');
+const undoDeathReportForm = require('../../../page-objects/default/enketo/undo-death-report.page');
+
+describe('Mute/Unmute contacts using a specific form.', () => {
+  const places = placeFactory.generateHierarchy();
+  const healthCenter = places.get('health_center');
+  const offlineUser = userFactory.build({ place: healthCenter._id, roles: ['chw'] });
+  const person = personFactory.build({ parent: {_id: healthCenter._id, parent: healthCenter.parent} });
+  const mutePerson = personFactory.build({ 
+    patient_id: 12345,
```
Comment (lorerod) [inline tests/e2e/default/contacts/muting-contact-using-form.wdio-spec.js:19]: Is this patient_id value really need to be set? or you can use the default value set in `personFactory`

File: tests/e2e/default/contacts/muting-contact-using-form.wdio-spec.js [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    await genericForm.submitForm();
+    await commonPage.waitForPageLoaded();
+    await commonPage.sync(true);
+
+    expect(await (await contactPage.contactMuted()).isDisplayed()).to.be.false;
+
+  });
+
+  it('Should show a popup when trying to submit a non-unmuting form against a muted contact', async () => {
+    await utils.revertSettings(true);
```
Comment (lorerod) [inline tests/e2e/default/contacts/muting-contact-using-form.wdio-spec.js:66]: Why do we need to revert the settings? this mean the `setting_mute` setting is not going to be active for this scenario?

**Review by lorerod — APPROVED** [review]
Thank you, @tatilepizs, for this e2e test. It took you a lot of work to finish this in between all the other stuff you had going on. It looks great! Very tidy and clear. I just left a couple of questions.

File: tests/e2e/default/contacts/muting-contact-using-form.wdio-spec.js (hunk 2 above)
Comment (tatilepizs) [reply tests/e2e/default/contacts/muting-contact-using-form.wdio-spec.js:19]: I set a different value for the `patient_id` because the `person` was already using the default value that was assigned in the `personFactory`

File: tests/e2e/default/contacts/muting-contact-using-form.wdio-spec.js (hunk 3 above)
Comment (tatilepizs) [reply tests/e2e/default/contacts/muting-contact-using-form.wdio-spec.js:66]: Yes, because what we are trying to check here is that if we have a muted person, and we want to submit a form,  a form that is not configured in the muting settings, a pop-up should be displayed saying that you are trying to submit a form against a muted person.

File: tests/e2e/default/contacts/muting-contact-using-form.wdio-spec.js (hunk 2 above)
Comment (lorerod) [reply tests/e2e/default/contacts/muting-contact-using-form.wdio-spec.js:19]: Oh! ok I get it. Thanks.

File: tests/e2e/default/contacts/muting-contact-using-form.wdio-spec.js (hunk 3 above)
Comment (lorerod) [reply tests/e2e/default/contacts/muting-contact-using-form.wdio-spec.js:66]: I want to make sure I understand. We will have the same results keeping the `setting_mute` setting active, and using a form, for example, pregnancy, unrelated to the `setting_mute`?

Comment (tatilepizs) [reply tests/e2e/default/contacts/muting-contact-using-form.wdio-spec.js:66]: That is correct.
The forms that are "affected" by the `muting` setting are the ones assigned to `mute_forms` or `unmute_forms`.
I can show you how to test that manually, maybe that can help. It was also hard for me to understand it at the beginning, I had to do a lot of different tests. 
[Here](https://docs.communityhealthtoolkit.org/apps/reference/app-settings/transitions/#muting) is more information

Comment (lorerod) [reply tests/e2e/default/contacts/muting-contact-using-form.wdio-spec.js:66]: Thanks @tatilepizs I think it is clearer now. It is a really interesting feature to try edge cases in the future. Great job!

Comment (tatilepizs) [reply tests/e2e/default/contacts/muting-contact-using-form.wdio-spec.js:66]: Thank you! 😊
